pkg/tiup/deploy.go

```diff
@@ -579,3 +691,158 @@ func (c *ClusterManager) GenUnzipCommand(filePath string, destDir string) string
 	}
 	return "tar -xf " + filePath + " -C " + destDir
 }
+
+func (c *ClusterManager) getScaleInNode(pre naglfarv1.TiDBCluster, cur naglfarv1.TiDBCluster, clusterIPMaps map[string]string) []string {
+	var result []string
+	for i := 0; i < len(pre.TiDB); i++ {
+		var isExisted bool
+		for j := 0; j < len(cur.TiDB); j++ {
+			if reflect.DeepEqual(pre.TiDB[i], cur.TiDB[j]) {
+				isExisted = true
+			}
+		}
+		if !isExisted {
+			if pre.TiDB[i].Port == 0 {
+				pre.TiDB[i].Port = 4000
+			}
+			node := clusterIPMaps[pre.TiDB[i].Host] + ":" + strconv.Itoa(pre.TiDB[i].Port)
+			result = append(result, node)
+		}
+	}
+	for i := 0; i < len(pre.PD); i++ {
+		var isExisted bool
+		for j := 0; j < len(cur.PD); j++ {
+			if reflect.DeepEqual(pre.PD[i], cur.PD[j]) {
+				isExisted = true
+			}
+		}
+		if !isExisted {
+			if pre.PD[i].ClientPort == 0 {
+				pre.PD[i].ClientPort = 2379
+			}
+			node := clusterIPMaps[pre.PD[i].Host] + ":" + strconv.Itoa(pre.PD[i].ClientPort)
+			result = append(result, node)
+		}
+	}
+	for i := 0; i < len(pre.TiKV); i++ {
+		var isExisted bool
+		for j := 0; j < len(cur.TiKV); j++ {
+			if reflect.DeepEqual(pre.TiKV[i], cur.TiKV[j]) {
+				isExisted = true
+			}
+		}
+		if !isExisted {
+			if pre.TiKV[i].Port == 0 {
+				pre.TiKV[i].Port = 20160
+			}
+			node := clusterIPMaps[pre.TiKV[i].Host] + ":" + strconv.Itoa(pre.TiKV[i].Port)
+			result = append(result, node)
+		}
+	}
+	return result
+}
+
+func (c *ClusterManager) getScaleOutRoles(pre naglfarv1.TiDBCluster, cur naglfarv1.TiDBCluster, clusterIPMaps map[string]string) ([]naglfarv1.TiDBSpec, []naglfarv1.PDSpec, []naglfarv1.TiKVSpec) {
+
+	var tidbs []naglfarv1.TiDBSpec
+	for i := 0; i < len(cur.TiDB); i++ {
+		var isExisted bool
+		for j := 0; j < len(pre.TiDB); j++ {
+			if reflect.DeepEqual(cur.TiDB[i], pre.TiDB[j]) {
```

Comment:
Some details need attention:

How about if I modify spec like this:

```diff
        statusPort: 20180
        deployDir: /disk1/deploy/tikv-20160
        dataDir: /disk1/data/tikv-20160
---   logDir: /disk1/deploy/tikv-20160/log
+++   logDir: /disk1/deploy/tikv-20160/log2
```

Reply:
Yes, more stringent inspection is needed here.